src/core/net/dns_client.cpp

```diff
@@ -43,357 +43,778 @@
 {
-    otError  error       = OT_ERROR_NONE;
-    Message *messageCopy = aMessage.Clone();
+    // Search and read SRV, TXT records in Answer Section
+    // matching name from query.
 
-    VerifyOrExit(messageCopy != nullptr, error = OT_ERROR_NO_BUFS);
+    return FindServiceInfo(kAnswerSection, Name(*mQuery, kNameOffsetInQuery), aServiceInfo);
+}
 
-    SuccessOrExit(error = aQueryMetadata.AppendTo(*messageCopy));
-    mPendingQueries.Enqueue(*messageCopy);
+otError Client::ServiceResponse::GetHostAddress(const char *  aHostName,
+                                                uint16_t      aIndex,
+                                                Ip6::Address &aAddress,
+                                                uint32_t &    aTtl) const
+{
+    return FindHostAddress(kAdditionalDataSection, Name(aHostName), aIndex, aAddress, aTtl);
+}
 
-    mRetransmissionTimer.FireAtIfEarlier(aQueryMetadata.mTransmissionTime);
+#endif // OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+
+//---------------------------------------------------------------------------------------------------------------------
+// Client
+
+const uint16_t Client::kAddressQueryRecordTypes[] = {ResourceRecord::kTypeAaaa};
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+const uint16_t Client::kBrowseQueryRecordTypes[]  = {ResourceRecord::kTypePtr};
+const uint16_t Client::kServiceQueryRecordTypes[] = {ResourceRecord::kTypeSrv, ResourceRecord::kTypeTxt};
+#endif
+
+const uint8_t Client::kQuestionCount[] = {
+    /* (0) kAddressQuery -> */ OT_ARRAY_LENGTH(kAddressQueryRecordTypes), // AAAA records
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+    /* (1) kBrowseQuery  -> */ OT_ARRAY_LENGTH(kBrowseQueryRecordTypes),  // PTR records
+    /* (2) kServiceQuery -> */ OT_ARRAY_LENGTH(kServiceQueryRecordTypes), // SRV and TXT records
+#endif
+};
+
+const uint16_t *Client::kQuestionRecordTypes[] = {
+    /* (0) kAddressQuery -> */ kAddressQueryRecordTypes,
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+    /* (1) kBrowseQuery  -> */ kBrowseQueryRecordTypes,
+    /* (2) kServiceQuery -> */ kServiceQueryRecordTypes,
+#endif
+};
+
+Client::Client(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mSocket(aInstance)
+    , mTimer(aInstance, Client::HandleTimer, this)
+{
+    static_assert(kAddressQuery == 0, "kAddressQuery value is not correct");
+#if OPENTHREAD_CONFIG_DNS_CLIENT_SERVICE_DISCOVERY_ENABLE
+    static_assert(kBrowseQuery == 1, "kBrowseQuery value is not correct");
+    static_assert(kServiceQuery == 2, "kServiceQuery value is not correct");
+#endif
+}
+
+otError Client::Start(void)
+{
+    otError error;
+
+    SuccessOrExit(error = mSocket.Open(&Client::HandleUdpReceive, this));
+    SuccessOrExit(error = mSocket.Bind());
 
 exit:
-    FreeAndNullMessageOnError(messageCopy, error);
-    return messageCopy;
+    return error;
 }
 
-void Client::DequeueMessage(Message &aMessage)
+otError Client::Stop(void)
```

Comment:
nit: we should always be able to stop the client?
```suggestion
void Client::Stop(void)
```